File: MyMusic.Mobile/src/services/sync/atomic-operations.ts
import type { ISyncApiClient, IFileOps, IUserPrompt, RecordItem, SyncContext, SyncFileBase, SyncRecordAction } from './types';
import { safeToIsoString } from './utils';

export async function uploadOneFile (
    apiClient: ISyncApiClient,
    ctx: SyncContext,
    file: SyncFileBase,
    action: SyncRecordAction,
    reason?: string
): Promise<RecordItem> {
    if (ctx.options.dryRun) {
        ctx.result[action === 'Created' ? 'created' : 'updated']++;
        return {
            filePath: file.relativePath,
            action,
            source: 'Device',
            reason,
        };
    }

    try {
        const filename = file.relativePath.split('/').pop();
        if (!filename) {
            throw new Error(`Invalid file path: relativePath is empty for ${file.fullPath}`);
        }

        await apiClient.uploadFile(
            ctx.deviceId,
            {
                uri: file.fullPath,
                name: filename,
            },
            file.relativePath,
            safeToIsoString(file.modifiedAt)!,
            safeToIsoString(file.createdAt)!
        );
        ctx.result[action === 'Created' ? 'created' : 'updated']++;
        return {
            filePath: file.relativePath,
            action,
            source: 'Device',
            reason,
        };
    } catch (e) {
        ctx.result.failed++;
        const errorMessage = e instanceof Error ? e.message : String(e);
        return {
            filePath: file.relativePath,
            action: 'Error',
            source: 'Device',
            reason,
            errorMessage,
        };
    }
}

export async function downloadOneFile (
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    ctx: SyncContext,
    songId: number,
    path: string,
    decodedRepoPath: string,
    previousPath?: string | null
): Promise<RecordItem | null> {
    const fullPath = `${decodedRepoPath}/${path}`;
    const previousFullPath = previousPath ? `${decodedRepoPath}/${previousPath}` : null;

    if (ctx.options.dryRun) {
        ctx.result.downloaded++;
        return {
            filePath: path,
            action: 'Downloaded',
            source: 'Server',
            songId,
            reason: 'Server-initiated download',
        };
    }

    try {
        await fileOps.ensureDirectory(fullPath);

        if (fileOps.fileExists(fullPath)) {
            await fileOps.deleteFile(fullPath);
        }

        const blob = await apiClient.downloadSong(songId);
        await fileOps.writeFile(fullPath, blob);

        if (previousFullPath && fileOps.fileExists(previousFullPath)) {
            await fileOps.deleteFile(previousFullPath);
            await fileOps.deleteEmptyDirectories(previousFullPath, decodedRepoPath);
        }

        const modifiedAt = fileOps.getModificationTime(fullPath);
        await apiClient.acknowledgeAction(ctx.deviceId, {
            devicePath: path,
            modifiedAt: modifiedAt ? safeToIsoString(modifiedAt) : undefined,
            previousDevicePath: previousPath,
        });

        ctx.result.downloaded++;
        return {
            filePath: path,
            action: 'Downloaded',
            source: 'Server',
            songId,
            reason: previousPath ? 'Server-initiated rename' : 'Server-initiated download',
        };
    } catch (e) {
        ctx.result.failed++;
        const errorMessage = e instanceof Error ? e.message : String(e);
        return {
            filePath: path,
            action: 'Error',
            source: 'Server',
            songId,
            reason: 'Server-initiated download failed',
            errorMessage,
        };
    }
}

export async function removeOneFile (
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    userPrompt: IUserPrompt,
    ctx: SyncContext,
    path: string,
    decodedRepoPath: string
): Promise<RecordItem | null> {
    const fullPath = `${decodedRepoPath}/${path}`;

    if (!fileOps.fileExists(fullPath)) {
        if (!ctx.options.dryRun) {
            await apiClient.acknowledgeAction(ctx.deviceId, { devicePath: path });
        }
        return null;
    }

    let shouldDelete = true;

    if (!ctx.options.autoConfirm && !ctx.options.dryRun) {
        shouldDelete = await userPrompt.confirmDeletion(path);
    }

    if (!shouldDelete) {
        return null;
    }

    if (!ctx.options.dryRun) {
        await fileOps.deleteFile(fullPath);
        await apiClient.acknowledgeAction(ctx.deviceId, { devicePath: path });
    }

    ctx.result.removed++;
    return {
        filePath: path,
        action: 'Removed',
        source: 'Server',
        reason: 'Server-initiated removal',
    };
}

File: MyMusic.CLI/Services/Sync/AtomicOperations.cs
namespace MyMusic.CLI.Services.Sync;

using System.IO.Abstractions;
using Microsoft.Extensions.Logging;
using MyMusic.CLI.Services.Sync;
using MyMusic.CLI.Services.Sync.Types;

public class AtomicOperations(
    IFileSystem fileSystem,
    ISyncApiClient apiClient,
    IFileOps fileOps,
    IUserPrompt userPrompt,
    ILogger<AtomicOperations> logger)
{
    public async Task<RecordItem> UploadOneFileAsync(
        long deviceId,
        string repositoryPath,
        SyncFileInfo fileInfo,
        bool dryRun,
        CancellationToken ct = default)
    {
        var fullPath = Path.Combine(repositoryPath, fileInfo.Path);
        if (!fileSystem.File.Exists(fullPath))
        {
            logger.LogWarning("File not found: {Path}", fullPath);
            return new RecordItem
            {
                FilePath = fileInfo.Path,
                Action = "Error",
                Source = "Device",
                Reason = "File not found"
            };
        }

        if (dryRun)
        {
            return new RecordItem
            {
                FilePath = fileInfo.Path,
                Action = "Created",
                Source = "Device",
                Reason = fileInfo.Reason
            };
        }

        try
        {
            await using var stream = fileSystem.File.OpenRead(fullPath);
            var fileName = Path.GetFileName(fullPath);
            await apiClient.UploadFileAsync(deviceId, new UploadFileRequest
            {
                FileStream = stream,
                FileName = fileName,
                Path = fileInfo.Path,
                ModifiedAt = fileInfo.ModifiedAt.ToString("O"),
                CreatedAt = fileInfo.CreatedAt.ToString("O")
            }, ct);

            return new RecordItem
            {
                FilePath = fileInfo.Path,
                Action = "Created",
                Source = "Device",
                Reason = fileInfo.Reason
            };
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to upload file: {Path}", fileInfo.Path);
            return new RecordItem
            {
                FilePath = fileInfo.Path,
                Action = "Error",
                ErrorMessage = ExtractErrorMessage(ex),
                Source = "Device",
                Reason = fileInfo.Reason
            };
        }
    }

    public async Task<RecordItem?> DownloadOneFileAsync(
        long songId,
        string fullPath,
        string relativePath,
        bool dryRun,
        bool autoConfirm,
        CancellationToken ct = default)
    {
        var fileExists = fileOps.FileExists(fullPath);

        if (!dryRun && fileExists && !autoConfirm)
        {
            var confirmed = await userPrompt.ConfirmDeletionAsync(relativePath, ct);
            if (!confirmed)
            {
                return null;
            }
        }

        if (dryRun)
        {
            return new RecordItem
            {
                FilePath = relativePath,
                Action = "Downloaded",
                SongId = songId,
                Source = "Server",
                Reason = "Server-initiated download"
            };
        }

        var tempPath = fullPath + ".tmp";
        try
        {
            await fileOps.EnsureDirectoryAsync(fullPath, ct);

            await using var stream = await apiClient.DownloadSongAsync(songId, ct);
            await fileOps.WriteFileAsync(tempPath, stream, ct);

            if (fileExists)
            {
                await fileOps.DeleteFileAsync(fullPath, ct);
            }

            fileSystem.File.Move(tempPath, fullPath);

            var modifiedAt = await fileOps.GetModificationTimeAsync(fullPath, ct);
            return new RecordItem
            {
                FilePath = relativePath,
                Action = "Downloaded",
                SongId = songId,
                Source = "Server",
                Reason = "Server-initiated download"
            };
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to download file: {Path}", relativePath);
            return new RecordItem
            {
                FilePath = relativePath,
                Action = "Error",
                SongId = songId,
                ErrorMessage = ex.Message,
                Source = "Server",
                Reason = "Server-initiated download failed"
            };
        }
        finally
        {
            if (fileSystem.File.Exists(tempPath))
            {
                fileSystem.File.Delete(tempPath);
            }
        }
    }

    public async Task<RecordItem?> RemoveOneFileAsync(
        long deviceId,
        string fullPath,
        string relativePath,
        bool dryRun,
        bool autoConfirm,
        CancellationToken ct = default)
    {
        var fileExists = fileOps.FileExists(fullPath);

        if (!fileExists)
        {
            if (!dryRun)
            {
                await apiClient.AcknowledgeActionAsync(deviceId, new AcknowledgeActionRequest
                {
                    DevicePath = relativePath
                }, ct);
            }
            return null;
        }

        if (!dryRun && !autoConfirm)
        {
            var confirmed = await userPrompt.ConfirmDeletionAsync(relativePath, ct);
            if (!confirmed)
            {
                return null;
            }
        }

        if (dryRun)
        {
            return new RecordItem
            {
                FilePath = relativePath,
                Action = "Removed",
                Source = "Server",
                Reason = "Server-initiated removal"
            };
        }

        try
        {
            await fileOps.DeleteFileAsync(fullPath, ct);
            await apiClient.AcknowledgeActionAsync(deviceId, new AcknowledgeActionRequest
            {
                DevicePath = relativePath
            }, ct);

            return new RecordItem
            {
                FilePath = relativePath,
                Action = "Removed",
                Source = "Server",
                Reason = "Server-initiated removal"
            };
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to delete file: {Path}", relativePath);
            return new RecordItem
            {
                FilePath = relativePath,
                Action = "Error",
                ErrorMessage = ex.Message,
                Source = "Server",
                Reason = "Server-initiated removal failed"
            };
        }
    }

    private static string ExtractErrorMessage(Exception ex)
    {
        return ex.Message;
    }
}
